pulsar-broker/src/main/java/org/apache/pulsar/broker/web/RestEndpointMetricsFilter.java

```diff
@@ -0,0 +1,106 @@
+import org.glassfish.jersey.server.model.ResourceMethod;
+import org.glassfish.jersey.uri.UriTemplate;
+
+public class RestEndpointMetricsFilter implements ContainerResponseFilter, ContainerRequestFilter {
+    private static final String REQUEST_START_TIME = "requestStartTime";
+    private static final AttributeKey<String> PATH = SemanticAttributes.URL_PATH;
+    private static final AttributeKey<String> METHOD = SemanticAttributes.HTTP_REQUEST_METHOD;
+    private static final AttributeKey<Long> CODE = SemanticAttributes.HTTP_RESPONSE_STATUS_CODE;
+
+    private final DoubleHistogram latency;
+
+    private RestEndpointMetricsFilter(PulsarBrokerOpenTelemetry openTelemetry) {
+        Meter meter = openTelemetry.getMeter();
+        latency = meter.histogramBuilder("pulsar_broker_rest_endpoint_latency")
+                .setDescription("Latency of REST endpoints in Pulsar broker")
+                .setUnit("ms")
+                .setExplicitBucketBoundariesAdvice(List.of(10D, 20D, 50D, 100D, 200D, 500D, 1000D, 2000D))
+                .build();
+    }
+
+    private static volatile RestEndpointMetricsFilter instance;
+
+    public static synchronized RestEndpointMetricsFilter create(PulsarBrokerOpenTelemetry openTelemetry) {
+        if (instance == null) {
+            instance = new RestEndpointMetricsFilter(openTelemetry);
+        }
+        return instance;
+    }
+
+    @Override
+    public void filter(ContainerRequestContext req, ContainerResponseContext resp) throws IOException {
+        Response.StatusType status = resp.getStatusInfo();
+        int statusCode = status.getStatusCode();
+        Attributes attrs;
+        try {
+            UriRoutingContext info = (UriRoutingContext) req.getUriInfo();
+            attrs = getRequestAttributes(info, statusCode);
+        } catch (Throwable ex) {
+            attrs = Attributes.of(PATH, "UNKNOWN", METHOD, req.getMethod(), CODE, (long) statusCode);
+        }
+
+        Object o = req.getProperty(REQUEST_START_TIME);
+        if (o instanceof Long start) {
+            long duration = System.currentTimeMillis() - start;
+            this.latency.record(duration, attrs);
+        }
+    }
+
+    @Override
+    public void filter(ContainerRequestContext req) throws IOException {
+        // Set the request start time into properties.
+        req.setProperty(REQUEST_START_TIME, System.currentTimeMillis());
+    }
+
+    private static Attributes getRequestAttributes(UriRoutingContext ctx, long statusCode) {
+        List<UriTemplate> templates = ctx.getMatchedTemplates();
+        ResourceMethod method = ctx.getMatchedResourceMethod();
+        String httpMethod = method == null ? "UNKNOWN" : method.getHttpMethod();
+        if (CollectionUtils.isEmpty(templates)) {
+            return Attributes.of(PATH, "UNKNOWN", METHOD, httpMethod, CODE, statusCode);
+        }
+        UriTemplate[] arr = templates.toArray(new UriTemplate[0]);
```

Comment:
Why do you specifically need to convert this to an array? Can't you just iterate over the list using `get(i)`?